inc/hclib-async.h

```diff
@@ -55,37 +56,69 @@ namespace hclib {
  * TODO optimize that overhead.
  */
 
+/* raw function pointer for calling lambdas */
 template<typename T>
 void lambda_wrapper(void *arg) {
-    T* lambda = static_cast<T*>(arg);
+    T *lambda = static_cast<T*>(arg);
     MARK_BUSY(current_ws()->id);
     (*lambda)(); // !!! May cause a worker-swap !!!
     MARK_OVH(current_ws()->id);
     delete lambda;
 }
 
-template <typename T>
-inline hclib_task_t* _allocate_async_hclib(T lambda, bool await) {
-    typedef typename std::remove_reference<T>::type U;
-    // FIXME - this whole call chain is kind of a mess
-    // leaving C malloc/free and memcpy calls for now (come back to fix it later)
-    const size_t task_size =
-        await ? sizeof(hclib_dependent_task_t) : sizeof(hclib_task_t);
-    hclib_task_t* task = static_cast<hclib_task_t*>(calloc(1, task_size));
-    task->_fp = lambda_wrapper<U>;
-
-    // make a copy of the user's function object using dynamic storage duration
-    // to ensure that it's still available later.
-    task->args = new U(lambda);
-
-    return task;
-}
 
-#if defined(HUPCPP) && defined(DIST_WS)	// i.e. if we are supporting distributed work-stealing in HabaneroUPC++
-#define _allocate_async _allocate_async_hcupc
-#else
-#define _allocate_async _allocate_async_hclib
-#endif
+/* this version also deletes a runtime-managed future list */
+template<typename T>
+struct LambdaAwaitArgs {
```

Comment:
This change is part of #20. That's a good point though. I do prefer PascalCase for type names (in both C and C++), since that makes it easy to avoid name conflicts between types and data members; however, consistency obviously trumps my personal preferences, so I'll go back and rebase this to use snake_case instead.